### PR TITLE
Fix DTensor backward for value-selecting reductions (topk, sort, min,…

### PR DESCRIPTION
**Pull request by dcci**
… max)

`value_selecting_reduction_backward` creates a zeros tensor via `at::zeros_symint(sizes, grad_out.options())` which bypasses tensor subclass dispatch since `options()` only carries dtype/device. When `grad_out` is a DTensor, the resulting plain zeros tensor is passed to `scatter` alongside DTensor arguments, causing a mixed-type error.

Fix: use `grad_out.new_zeros_symint(sizes)` in the tensor subclass path so that subclasses (DTensor, FakeTensor, etc.) can intercept the creation through dispatch and produce a properly-typed result.

Authored with Claude

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]

File: test/distributed/tensor/test_math_ops.py [hunk 1]
```diff
@@ -688,9 +688,34 @@ def test_topk(self):
...
+                12, 8, 8, device=self.device_type, requires_grad=True
+            )
+            dist_x = DTensor.from_local(local_x, device_mesh, [Shard(shard_dim)])
+            dist_x.retain_grad()
+
+            dist_topk = dist_x.topk(3, dim=topk_dim)
+            dist_loss = dist_topk.values.sum()
+            dist_loss.backward()
+
+            # Verify gradient exists and has the right placement
```
Comment (stmcgovern) [inline test/distributed/tensor/test_math_ops.py]: Should the test check to see if the gradient is correct?

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -2351,10 +2351,14 @@ bool cpu_equal(const Tensor& self, const Tensor& other) {
 Tensor value_selecting_reduction_backward_symint(const Tensor& grad, int64_t dim, const Tensor& indices, c10::SymIntArrayRef sizes, bool keepdim) {
   auto inplace_scatter_if_not_tensor_subclass =
       [&](const Tensor& grad_out, const Tensor& indices_) {
-        auto grad_in = at::zeros_symint(sizes, grad_out.options());
         if (areAnyTensorSubclassLike({grad, indices})) {
+          // Use new_zeros_symint so that tensor subclasses (e.g. DTensor)
+          // can intercept the zeros creation through dispatch, ensuring
+          // the result has matching subclass type for subsequent scatter.
+          auto grad_in = grad_out.new_zeros_symint(sizes);
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp:2358]: Is there no reason not to ask Claude to intercept and fix any issues like this atleast in a similar if statement dispatch?

Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp:2358]: Looks like cummaxmin_backward will have the same backward issue for instance...

Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp:2358]: Last why not use the zeros_symint dispatch unconditionally?

Comment (dcci) [reply aten/src/ATen/native/ReduceOps.cpp:2358]: The overhead is too high

Comment (dcci) [reply aten/src/ATen/native/ReduceOps.cpp:2358]: we can have cummaxmin as follow up

File: test/distributed/tensor/test_math_ops.py (hunk 1 above)
Comment (dcci) [reply test/distributed/tensor/test_math_ops.py]: yes, good idea, let me add it
